### PR TITLE
feat: classification chain

### LINKED ISSUE #61: [FEATURE] Enable chat setting for all users
### Describe the feature

Now only Admin user can configure chat setting, but non-admin users also want to test various models and do prompt engineering.

### Use Case

Non-admin users are disappointed in not configuring chat setting.

### Proposed Solution

Enable non-admin users to configure chat setting.

### Other Information

_No response_

### Acknowledgements

- [ ] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### PR DESCRIPTION
**Pull request by JeremyJonas**
## Description

We now support enabling a pre-processing step to the conversational chain, called "classify", which is very flexible and powerful way to provide additional functionality and control over your chain. The classify sends the user question as input to LLM and returns JSON data which is parsed and provided to the following prompts. This can be used to detect original language, to perform translation, to categorize the question based on your own categorize, etc. This is very powerful when joined with the handlebar templates which can dynamically modify your prompts based on the output from classify step. Think prompt template selector but in a single handlebars template.

### Bug Fixes

* **adapter/claude:** use alternal message prefix ([cd5287c](https://github.com/aws-samples/aws-genai-conversational-rag-reference/commit/cd5287c5ea8b13c9723a7746fa1b9724fc5039d4))
* **cli:** cognito prompts ([582a40b](https://github.com/aws-samples/aws-genai-conversational-rag-reference/commit/582a40bfa26a97d5890f35f04590bed3d4afcbcc))
* **cli:** resolve cache parse issue and exit issues ([5a4ed49](https://github.com/aws-samples/aws-genai-conversational-rag-reference/commit/5a4ed49906916242aba1e6b3603cadd4480060f9))
* **cli:** support doc metadata values other than string ([b827090](https://github.com/aws-samples/aws-genai-conversational-rag-reference/commit/b8270907f4f66b1e79db8b2590abfd8e78ba0020))
* **corpus:** support metadata with non-string values ([e389a0e](https://github.com/aws-samples/aws-genai-conversational-rag-reference/commit/e389a0ebfd3a7137e0a12a55774db852967b8bb1))
* **infra/website:** resolve timeout and broken ux for delete chat ([74ce834](https://github.com/aws-samples/aws-genai-conversational-rag-reference/commit/74ce8340e01e89073a4d345c25ef6ea716a2dc26))
* **metrics:** correct metric publishing and service naming ([59e5cb2](https://github.com/aws-samples/aws-genai-conversational-rag-reference/commit/59e5cb2963084111b0d022f8f815df908b32ece5))
* **sdk:** use api types in sdk to considate/simplify chat engine config ([622736c](https://github.com/aws-samples/aws-genai-conversational-rag-reference/commit/622736cfea96aa3ab55ea7eeecbdcb9b2014e543))
* **website:** browserify node natives from sdk ([99074b5](https://github.com/aws-samples/aws-genai-conversational-rag-reference/commit/99074b5d01836361be8f03f698395303663d560e))
* **website:** chat-engine-config override copy task ([b373d3d](https://github.com/aws-samples/aws-genai-conversational-rag-reference/commit/b373d3d7cf32c98c474a685cce49a8506b3b519e))
* **website:** fix parsing of chat setting model/endpoint kwargs ([856f195](https://github.com/aws-samples/aws-genai-conversational-rag-reference/commit/856f19583ff3e8c3d1dad44a2a85f489b1d2a34a))
* **website:** fix prompt handlebars message rendering ([065373f](https://github.com/aws-samples/aws-genai-conversational-rag-reference/commit/065373f0fa2754c6afbf9c6c7a73d9d49c68a93e))
* **website:** prevent continuous chat input updates ([3332ba2](https://github.com/aws-samples/aws-genai-conversational-rag-reference/commit/3332ba256172905b12069101717e7fa27c815bec))
* **website:** resolve scroll to bottom on new message ([66230d3](https://github.com/aws-samples/aws-genai-conversational-rag-reference/commit/66230d38db03ed605aeea4fea7b69f8425a78e31))

### Features

* **chat:** add query execution time metric ([8b1021b](https://github.com/aws-samples/aws-genai-conversational-rag-reference/commit/8b1021bf5a4e075812a6a5d5b6b21cc4a943a743))
* **chat:** enable handlebars control over context documents in prompt ([1bd907d](https://github.com/aws-samples/aws-genai-conversational-rag-reference/commit/1bd907d19aa5551b51a575a78bd6bc7f8fa23719))
* **cli:** add `invoke chat-bulk` command ([9075e20](https://github.com/aws-samples/aws-genai-conversational-rag-reference/commit/9075e20ae3e26d91a697e758246e063d549f811b))
* **cli:** add cognito auth command ([c50e688](https://github.com/aws-samples/aws-genai-conversational-rag-reference/commit/c50e688b95e959a754beaf23f6528a02078ae772))
* **cli:** chat-bulk parallelization ([d158d56](https://github.com/aws-samples/aws-genai-conversational-rag-reference/commit/d158d569160a30d73c5096a51d686954fa570dd6))
* **cli:** make cognito user pool selection cached ([b3dfebc](https://github.com/aws-samples/aws-genai-conversational-rag-reference/commit/b3dfebcc9c81dc97c7b47cc95519f840fe592b51))
* **demo/overrides:** add chat-engine-config overrides ([f4d008e](https://github.com/aws-samples/aws-genai-conversational-rag-reference/commit/f4d008e72b66af8f6def9d543d8571b166ce6b8f))
* **infra/corpus:** make postgres index creation "opt-in" ([2cad182](https://github.com/aws-samples/aws-genai-conversational-rag-reference/commit/2cad182fefc4822b2fc6fb7ba74b5af1346ee387))
* major chat engine config overhaul + classify chain ([ca4670e](https://github.com/aws-samples/aws-genai-conversational-rag-reference/commit/ca4670e87dd6022a806f4c1c0fb868035c7388ef)), closes [#61](https://github.com/aws-samples/aws-genai-conversational-rag-reference/issues/61)
* **prompt/adapter:** simplify prompt adapter ([e7bf225](https://github.com/aws-samples/aws-genai-conversational-rag-reference/commit/e7bf2253926ef0c9d7f942d1b3398349d8b25658))
* **sdk:** support browser logger variant with esm ([b2c775e](https://github.com/aws-samples/aws-genai-conversational-rag-reference/commit/b2c775e3dabcfe29e7a770173e2ab6bbc6dc3d97))
* **website:** add metadata expandable to source documents ([a96825c](https://github.com/aws-samples/aws-genai-conversational-rag-reference/commit/a96825c7857f136948eae72ca66767cc3542d89c))
* **website:** chat title update on enter ([a742717](https://github.com/aws-samples/aws-genai-conversational-rag-reference/commit/a74271796e7795e8d312875295540e3c4083f2e1))

### BREAKING CHANGES

* - Changed prompt template partials to remove extra abstractions,
now just template and couple partials. Non system templates that use the old partials will break.
- ChatEngineConfig is restructured with non-compatible interface.
Previous configs will not work and must be reconstructed via UI or manually via json.
- UI Chat Settings actions "clear" and "reset" consolidated to just "reset",
which resets the config for the chat to default.
- UI Chat Settings is now available to all users but with privileged fields
such as custom models being unavailable. The engine handle will throw
an error if non-admin sends privileged keys

### CODE REVIEW

File: docs/content/development/dev-settings/prompting/classify/index.md [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+- Detect question language
+- Translate the question
+- Categorize the question
+- Sentiment analysis
+- ... whatever your LLM can do
+
+!!! tip "Multi-Language Support"
+
+    If your knowledge base is all in english but you want to support end-users in different language, you can translate the input question to english and capture the original language, using the english translation throughout the chain, and tell the final QA LLM to reply in the original language. This has been with Bedrock Claude v2 with great success, you automatically support all languages that the model supports.
```
Comment (sperka) [inline docs/content/development/dev-settings/prompting/classify/index.md]: This has been `tested` with Bedrock Claude

File: demo/infra/src/application/ai/inference-engine/engine/handler/index.ts [hunk 2]
```diff
@@ -12,6 +17,8 @@ import {
   ApiResponse,
 } from 'api-typescript-interceptors';
 import { CreateChatMessageResponseContent, createChatMessageHandler } from 'api-typescript-runtime';
+import { cloneDeepWith, isUndefined, omitBy } from 'lodash';
+import applicationChatEngineConfigJson from './chat-engine-config.json'; // HACK: temporary way to support updating app level config at deploy time
```
Comment (sperka) [inline demo/infra/src/application/ai/inference-engine/engine/handler/index.ts:21]: this makes build to fail if file is not present

**Review by sperka — CHANGES_REQUESTED** [review]
thanks for the docs! awesome!

File: demo/infra/src/application/ai/inference-engine/engine/handler/index.ts (hunk 2 above)
Comment (JeremyJonas) [reply demo/infra/src/application/ai/inference-engine/engine/handler/index.ts:21]: `pre-compile` task in projen was mis-configured - I have fixed it now and rebuilding

Comment (JeremyJonas) [reply demo/infra/src/application/ai/inference-engine/engine/handler/index.ts:21]: passed build :D

**Review by sperka — DISMISSED** [review]
🚀

**Review by martenpayne — COMMENTED** [review]
…review is WIP, having some build issues atm…

File: demo/website/src/providers/ChatEngineConfig.tsx [hunk 3]
```diff
@@ -1,20 +1,39 @@
 /*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
 PDX-License-Identifier: Apache-2.0 */
+// import { getChainConfigKeyByType } from '@aws/galileo-sdk/lib/chat/config';
```
Comment (martenpayne) [inline demo/website/src/providers/ChatEngineConfig.tsx]: dead code

File: packages/galileo-cli/src/commands/invoke/chat-bulk/index.ts [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+import { helpers } from '../../../internals';
+import { createSignedFetcher } from '../../../lib/account-utils/signed-fetch';
+import context from '../../../lib/context';
+import galileoPrompts from '../../../lib/prompts';
+import CognitoAuthCommand from '../../cognito/auth';
+
+const TOKEN_BUFFER = 300 * 60; // 5 minutes
+
+const LAMBDA_URL_PATTERN = /^https:\/\/(?<id>[^.]+)\.lambda-url\.(?<region>[^.]+)\.on\.aws/;
+
```
Comment (martenpayne) [inline packages/galileo-cli/src/commands/invoke/chat-bulk/index.ts:23]: might be worth an overview comment for this functionality

File: packages/galileo-sdk/src/chat/chain.ts [hunk 5]
```diff
@@ -2,13 +2,15 @@
 PDX-License-Identifier: Apache-2.0 */
 import { BaseLanguageModel } from 'langchain/base_language';
 import { CallbackManagerForChainRun } from 'langchain/callbacks';
-import { BaseChain, ChainInputs, LLMChain, loadQAChain, QAChainParams } from 'langchain/chains';
+import { BaseChain, ChainInputs, LLMChain, QAChainParams, StuffDocumentsChain } from 'langchain/chains';
 import { PromptTemplate } from 'langchain/prompts';
```
Comment (martenpayne) [inline packages/galileo-sdk/src/chat/chain.ts:6]: dead code?

File: packages/galileo-sdk/src/chat/chain.ts [hunk 6]
```diff
@@ -184,3 +270,22 @@ export class ChatEngineChain extends BaseChain implements ChatEngineChainInput {
   //   throw new Error("Not implemented.");
```
Comment (martenpayne) [inline packages/galileo-sdk/src/chat/chain.ts]: dead code?

File: packages/galileo-sdk/src/chat/config/index.test.ts [hunk 7]
```diff
@@ -0,0 +1,159 @@
+/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
+PDX-License-Identifier: Apache-2.0 */
+// @ts-ignore
```
Comment (martenpayne) [inline packages/galileo-sdk/src/chat/config/index.test.ts]: mention why

File: packages/galileo-sdk/src/chat/config/index.ts [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+  options?: { verbose?: boolean },
+): Promise<ResolvedLLM> {
+  let modelInfo = defaultValues?.modelInfo;
+  if (llmConfig?.model) {
+    modelInfo = await ChatEngineContext.resolveModelInfo(llmConfig.model);
+  }
+
+  return ChatEngineContext.resolveLLM(modelInfo, { ...llmConfig, ...options });
+}
+
```
Comment (martenpayne) [inline packages/galileo-sdk/src/chat/config/index.ts:96]: maybe add an overview comment

File: projenrc/demo/index.ts [hunk 9]
```diff
@@ -25,6 +27,10 @@ export class Demo {
       options;
 
     const api = new Api({ monorepo, rootOutdir });
+    // BAD: framework taking dep on demo is bad, but we are planning to drastically simplify
```
Comment (martenpayne) [inline projenrc/demo/index.ts]: tweak adjective - perhaps "WILLFIX", "HACK" or "WORKAROUND"

**Review by martenpayne — APPROVED** [review]
Looks good

**Review by sperka — APPROVED** [review]

**Pull request merged**